### ISSUE TITLE
Authentication fails when username equates to 0

### ISSUE THREAD

**Issue by one-humble-cranberry** [verdaccio/verdaccio#628]
Completely debugged this - will probably pull request a fix.

A username of any number of 0's will fail to authenticate (for context, I have a user 000 in my LDAP)

Here's why:

https://github.com/verdaccio/verdaccio/blob/master/src/lib/auth.js#L87
```js
const groups = [ "000" ];
if (groups != null && groups != false) {
  // code does not reach here
}

// say what !?!,... 
[ ] != false // false
[ 1 ] != false // true
[ 0 ] != false // false - I know right?
[ "0" ] != false // false - because "0" == 0
[ "000" ] != false // false - because "000" == 0
```

In the future, please always avoid "==" and "!=" - there are many resources online that will explain this in detail, but they are a huge source of bugs like this, where, yes it is an edge case - but this cost me 3 days of development time to debug.

The fix should look something like
```js
if (groups != null && groups.length) { /* ... */ }
// or if you want to be more strict 
if (groups instanceof Array && groups.length) { /* ... */ }
// of if you want to use builtins
if (Array.isArray(groups) && groups.length) { /* ... */ }
// see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray
// or if you want to be explicit
if (Array.isArray(groups) && groups.length !== 0) { /* ... */ }
```

Personally not a fan of `Array.isArray` for security purposes, but it's all a preference.

**Comment by juanpicado** [comment]
Hi, thank for the detailed explanation, I will be happy to see a PR. However, the code was inherited by Sinopia as completely disaster and last year we have tried to bring good practices to verdaccio, it is still WIP so your contribution is very welcome.

**Comment by one-humble-cranberry** [comment]
I see, makes sense - legacy code is always a pain point - I would be happy to make the change.

I was thinking as well, there is probably a reason to check for falsey values, not just strictly arrays. To avoid creating more bugs from this fix - I would recommend something like this:

```js
if (groups != null && (groups != false || groups.length)) { /* ... */ }
```

**Comment by juanpicado** [comment]
> Personally not a fan of Array.isArray for security purposes, but it's all a preference.

I'd like to know your reasons.

> I was thinking as well, there is probably a reason to check for falsey values, not just strictly arrays. 

Many plugins are legacy, we need to be careful here. 

Just as a note, we have to add some unit test for such edge cases.

**Comment by one-humble-cranberry** [comment]
Yes yes! I was just looking at how to add testing for this - I will add the suite (I believe it should be in auth.spec.js?) Don't worry I won't submit any pull request without some form of testing!!

And yes that's what I was thinking - do you have a specification for what this is actually looking for? I will be sure to add comments about what is expected, and why certain cases are looked for.

To me it looks like it wants to move to the next plugin if the value is falsey in any way OR if an array is passed which is empty - which would allow any truthy values (true, 1, new Date()) and should allow an array of some length greater than 0. 

I would recommend deprecating this for a future minor/major release (not sure your semver practices) so it can be locked down by types (not sure, new Date should be allowed, but it is currently so we will of course keep it for the time being)

My reasoning for security is that most browsers don't freeze constructor functionality so this is possible:
```js
const old_is_array = Array.isArray
Array.isArray = function ( ) {
  // send arguments to server for analysis
  old_is_array.apply(this, arguments);
};
```

And given you have auth going through this, information which is sensitive might be leaked (even accidentally) So yes, it's an edge case, but I only like to work with objects which are immutable just to not have the worry. (Also keep in mind a rogue plugin could inject that code into the page

**Comment by one-humble-cranberry** [comment]
Um - not sure what to make of this - I forked, and ran `yarn` `yarn run test` - and received this output. What am I doing wrong.
<img width="581" alt="screen shot 2018-03-16 at 10 17 28 pm" src="https://user-images.githubusercontent.com/36286778/37550680-e08c19b8-2967-11e8-80a4-0476cdf2c7a3.png">

**Comment by juanpicado** [comment]
@one-humble-cranberry https://gitter.im/verdaccio/development let's talk there whether you have issues with the stack
